torchdata/stateful_dataloader/sampler.py

```diff
@@ -125,3 +127,86 @@ def __iter__(self):
                     batch = [0] * self.batch_size
             if idx_in_batch > 0:
                 yield batch[:idx_in_batch]
+
+
+class _StatefulDistributedSamplerIterator(Iterator[int], Stateful):
+    def __init__(self, sampler):
+        self.sampler = sampler
+
+        if self.sampler.shuffle:
+            # deterministically shuffle based on epoch and seed
+            g = torch.Generator()
+            g.manual_seed(self.sampler.seed + self.sampler.epoch)
+            indices = torch.randperm(len(self.sampler.dataset), generator=g).tolist()  # type: ignore[arg-type]
+        else:
+            indices = list(range(len(self.sampler.dataset)))  # type: ignore[arg-type]
+
+        if not self.sampler.drop_last:
+            # add extra samples to make it evenly divisible
+            padding_size = self.sampler.total_size - len(indices)
+            if padding_size <= len(indices):
+                indices += indices[:padding_size]
+            else:
+                indices += (indices * math.ceil(padding_size / len(indices)))[:padding_size]
+        else:
+            # remove tail of data to make it evenly divisible.
+            indices = indices[: self.sampler.total_size]
+        assert len(indices) == self.sampler.total_size
+
+        # subsample
+        indices = indices[self.sampler.rank : self.sampler.total_size : self.sampler.num_replicas]
+        assert len(indices) == self.sampler.num_samples
+
+        self.parent_iterator = iter(indices)
+        self.indices = list(self.parent_iterator)
+        self.current_index = 0
+
+    def __next__(self) -> int:
+        if self.sampler.next_yielded is not None:
+            self.current_index = self.sampler.next_yielded
+            self.sampler.yielded = self.sampler.next_yielded
+            self.sampler.next_yielded = None
+        if self.current_index >= len(self.indices):
+            raise StopIteration
+
+        val = self.indices[self.current_index]
+        self.current_index += 1
+        self.sampler.yielded += 1
+        return val
+
+    def state_dict(self) -> Dict[str, Any]:
+        return self.sampler.state_dict()
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        self.sampler.load_state_dict(state_dict)
```

Comment:
I don't think we need this both here and in the main sampler class, can we consolidate to have this in just one place?